app/controllers/my_facilities/drug_stocks_controller.rb

```diff
@@ -0,0 +1,85 @@
+class MyFacilities::DrugStocksController < AdminController
+  include Pagination
+  include MyFacilitiesFiltering
+
+  layout "my_facilities"
+
+  before_action :authorize_my_facilities
+  after_action :verify_authorization_attempted
+  before_action :set_facility, only: [:new, :create]
+  before_action :set_for_end_of_month, only: [:new, :create]
+  before_action :drug_stocks_enabled?
+
+  def index
+    @facilities = filter_facilities
+      .includes(facility_group: :protocol_drugs)
+      .where(protocol_drugs: {stock_tracked: true})
+  end
+
+  def new
+    session[:report_url_with_filters] ||= request.referer
+    drug_stock_list = DrugStock.latest_for_facility(@facility, @for_end_of_month)
+    @drug_stocks = drug_stock_list.each_with_object({}) { |drug_stock, acc|
+      acc[drug_stock.protocol_drug.id] = drug_stock
+    }
+  end
+
+  def create
+    drug_stocks = DrugStock.transaction do
+      drug_stocks_reported.map do |drug_stock|
+        DrugStock.create!(facility: @facility,
+                          user: current_admin,
+                          protocol_drug_id: drug_stock[:protocol_drug_id],
+                          received: drug_stock[:received].presence,
+                          in_stock: drug_stock[:in_stock].presence,
+                          for_end_of_month: @for_end_of_month)
+      end
+    end
+
+    report_url_with_filters = session[:report_url_with_filters]
+    session[:report_url_with_filters] = nil
+    if drug_stocks.empty?
+      redirect_to report_url_with_filters
+    elsif drug_stocks.all?(&:valid?)
+      redirect_to report_url_with_filters, notice: "Saved drug stocks"
+    elsif drug_stocks.any?(&:invalid?)
```

Comment:
`create!` fails immediately on invalid inputs, so we return a `400 Bad Request` instead of the response with this message. Need to keep `create` or catch the exception.

Reply:
Ah yes, missed handling tx failure. Handled now in 94ab4cd4fbf6ef4693db28c7858306aecdfcbe37.